datajunction-server/datajunction_server/database/queryrequest.py

```diff
@@ -406,3 +407,234 @@ async def to_versioned_query_request(
             ],
             "orderby": orders,
         }
+
+
+@dataclass(order=True)
+class VersionedNodeKey:
+    """
+    A versioned key for a node, used to identify it in query requests.
+    """
+
+    name: str
+    version: str | None = None
+
+    def __str__(self) -> str:
+        return f"{self.name}@{self.version}" if self.version else self.name
+
+    def __eq__(self, value):
+        if isinstance(value, VersionedNodeKey):
+            return self.name == value.name and self.version == value.version
+        elif isinstance(value, str):
+            node_key = self.parse(value)
+            return self.name == node_key.name and self.version == node_key.version
+        return False
+
+    def __hash__(self) -> int:
+        return hash((self.name, self.version))
+
+    @classmethod
+    def from_node(cls, node: Node) -> "VersionedNodeKey":
+        """
+        Creates a versioned node key from a Node object.
+        """
+        return cls(name=node.name, version=node.current_version)
+
+    @classmethod
+    def parse(cls, node_key: str) -> "VersionedNodeKey":
+        """
+        Parses the versioned node key into a dictionary.
+        """
+        if "@" in node_key:
+            node_key, version = node_key.split("@", 1)
+            return VersionedNodeKey(name=node_key, version=version)
+        return VersionedNodeKey(name=node_key, version=None)
+
+
+@dataclass
+class VersionedQueryKey:
+    """
+    A versioned query request encapsulates the logic to version nodes, dimensions,
```

Comment:
I wonder why you simply didn't take the all the `QueryRequestParams` here and picked these only?

Reply:
The rest of the query params don't actually require any additional logic to set up versioning. Parameters like `include_all_columns` can just be included directly, whereas something like `default.num_repair_orders` should be frozen with the current version, so that we're always requesting SQL for the right version.